### PR TITLE
fix(playwright): preserve reporter finalization errors

### PR DESCRIPTION
**Pull request by juan-fernandez**
### What does this PR do?

Preserves Playwright Test Optimization data when a custom reporter fails during `onEnd` or `onExit`.

The change records the original reporter error before Playwright propagates, swallows, or converts it into a final run status. Test events that completed successfully remain passed, while the suite, module, and session are finalized as failed with the reporter error. The final writer flush remains bounded, including when an automatic failure-screenshot upload is still pending.

Reporter failures are exposed differently across supported Playwright versions:

| Playwright version | How the error is captured |
| --- | --- |
| 1.18–1.37 (dd-trace v5 only) | Wrap the legacy reporter multiplexer and record an `onEnd` failure before rethrowing it. |
| 1.38–1.59 | Wrap each created reporter's `onEnd` and `onExit` callbacks, record the error, and preserve Playwright's original behavior. |
| 1.60+ | Playwright moved the runner into bundled entry points. We inject an internal reporter that observes Playwright's private reporter-error fallback during finalization, records the original error, and restores `console.error` afterward. |

### Motivation

Playwright may handle a reporter exception after all tests have completed. Before this change, Test Optimization could miss that late failure.

For example:

1. A test passes.
2. A custom reporter throws `Error('report upload failed')` in `onEnd` or `onExit`.
3. Playwright exits non-zero.
4. Datadog nevertheless submits the suite, module, and session as passed because finalization did not observe the reporter error.

After this change, the completed test event is still reported as passed, but its suite, module, and session are reported as failed with `report upload failed`. Playwright still exits non-zero.

There was also a data-loss boundary around screenshot uploads. A failed test may finish while its automatic failure screenshot is uploading. If a reporter then fails and that upload never responds, the test event can remain buffered and session finalization can wait indefinitely. The error-finalization path now cancels the pending upload, marks the test with the screenshot-upload error tag, exports the test and parent events, and completes within the existing final-flush deadline.

`integration-tests/ci-visibility/playwright-pending-upload-finalization.js` is a deterministic fixture for that second case. It publishes the same Playwright diagnostic-channel events as a worker, supplies a failed serialized test plus an automatic screenshot, then reports a custom reporter error while the integration test keeps the media endpoint pending. This verifies cancellation, event delivery, failed parent status, original error metadata, non-zero exit behavior, and bounded completion without relying on a timing-sensitive real browser run.

### Additional Notes

Stacked on #9790. PR 3 provides deferred test session trace handling and the bounded final-flush APIs used here.

The internal reporter in `packages/datadog-instrumentations/src/playwright-reporter.js` is only the Playwright 1.60+ compatibility bridge. It does not create Test Optimization events and does not replace the user's configured reporter. It identifies reporter failures during Playwright's final reporter loop and publishes the error back to the main instrumentation, which owns suite/module/session finalization.

Validation:

- Reporter `onEnd` and `onExit` failures across Playwright 1.18 (v5), 1.38, 1.60–1.61, and latest-supported paths
- Primitive and hostile reporter rejection values
- Immutable `console.error`, custom `Error.prepareStackTrace`, and repeated programmatic runs
- Pending screenshot-upload cancellation and bounded finalization
- Completed test event plus failed suite/module/session assertions
- `npm run lint -- --quiet`

### CODE REVIEW

**Comment by juan-fernandez** [comment]
@codex review

File: packages/datadog-instrumentations/src/playwright.js [hunk 1]
```diff
@@ -1444,13 +1479,59 @@ function runAllTestsWrapper (runAllTests, playwrightVersion) {
...
+    for (const reporter of reporters) {
+      if (reporter[kDdPlaywrightReporterInstrumented] || typeof reporter.onEnd !== 'function') continue
+
+      Object.defineProperty(reporter, kDdPlaywrightReporterInstrumented, { value: true })
+      const onEnd = reporter.onEnd
+      reporter.onEnd = async function () {
+        try {
+          return await onEnd.apply(this, arguments)
+        } catch (error) {
+          reporterError ||= error
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/playwright.js]: Fixed in fix(playwright): handle unusual reporter failures (https://github.com/DataDog/dd-trace-js/commit/d478d1de5).

File: packages/datadog-instrumentations/src/playwright.js [hunk 2]
```diff
@@ -1444,13 +1479,59 @@ function runAllTestsWrapper (runAllTests, playwrightVersion) {
...
+ */
+function reportersHook (reportersPackage) {
+  return shimmer.wrap(reportersPackage, 'createReporters', createReporters => async function () {
+    const reporters = await createReporters.apply(this, arguments)
+    for (const reporter of reporters) {
+      if (reporter[kDdPlaywrightReporterInstrumented] || typeof reporter.onEnd !== 'function') continue
+
+      Object.defineProperty(reporter, kDdPlaywrightReporterInstrumented, { value: true })
+      const onEnd = reporter.onEnd
+      reporter.onEnd = async function () {
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/playwright.js]: Fixed in fix(playwright): handle unusual reporter failures (https://github.com/DataDog/dd-trace-js/commit/d478d1de5).

**Comment by juan-fernandez** [comment]
@codex review

File: packages/datadog-instrumentations/src/playwright-reporter.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+   * @returns {void}
+   */
+  onExit () {
+    // eslint-disable-next-line no-console
+    if (console.error === this.consoleError) {
+      // eslint-disable-next-line no-console
+      console.error = this.originalConsoleError
+    }
+  }
+
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/playwright-reporter.js]: Fixed in [fix(playwright): cover reporter finalization edge cases](https://github.com/DataDog/dd-trace-js/commit/6b0296ed5). Console interception is now restored only after the complete Playwright reporter lifecycle, so later onExit failures are captured.

File: packages/datadog-plugin-playwright/src/index.js [hunk 4]
```diff
@@ -127,8 +131,21 @@ class PlaywrightPlugin extends CiPlugin {
...
     }) => {
+      if (error) {
+        this.#isFinalizingAfterError = true
+        for (const testSuiteSpan of this._testSuiteSpansByTestSuiteAbsolutePath.values()) {
+          testSuiteSpan.setTag(TEST_STATUS, 'fail')
+          testSuiteSpan.setTag('error', error)
+        }
+        for (const [finishTest, abortController] of this.#pendingTestFinishCallbacks) {
+          finishTest()
+          abortController.abort()
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-plugin-playwright/src/index.js:145]: Fixed in [fix(playwright): cover reporter finalization edge cases](https://github.com/DataDog/dd-trace-js/commit/6b0296ed5). Pending uploads now finalize with the screenshot upload-error result before aborting, preserving the error tag on the exported test.

File: packages/datadog-instrumentations/src/playwright-reporter.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+ *
+ * @param {unknown} message - First console.error argument
+ * @param {unknown} error - Second console.error argument
+ * @returns {boolean}
+ */
+function isPlaywrightReporterError (message, error) {
+  if (message !== PLAYWRIGHT_REPORTER_ERROR_MESSAGE || error == null) return false
+
+  const stack = new Error('Playwright reporter error provenance').stack
+  const caller = stack?.split('\n', 4)[3]
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/playwright-reporter.js]: Fixed in [fix(playwright): cover reporter finalization edge cases](https://github.com/DataDog/dd-trace-js/commit/6b0296ed5). Reporter-error provenance now temporarily uses V8 call sites and restores the user's stack formatter, avoiding assumptions about Error.prepareStackTrace's return type.

File: packages/datadog-instrumentations/src/playwright-reporter.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+/**
+ * Identifies the private Playwright 1.60-1.61 reporter error fallback without
+ * interpreting identical user console output as a framework error.
+ *
+ * @param {unknown} message - First console.error argument
+ * @param {unknown} error - Second console.error argument
+ * @returns {boolean}
+ */
+function isPlaywrightReporterError (message, error) {
+  if (message !== PLAYWRIGHT_REPORTER_ERROR_MESSAGE || error == null) return false
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/playwright-reporter.js]: Fixed in [fix(playwright): cover reporter finalization edge cases](https://github.com/DataDog/dd-trace-js/commit/6b0296ed5). Nullish reporter exceptions are now accepted once the console call is proven to originate from Playwright, with regressions for both null and undefined.

File: packages/datadog-instrumentations/src/playwright.js [hunk 7]
```diff
@@ -1444,13 +1482,64 @@ function runAllTestsWrapper (runAllTests, playwrightVersion) {
...
+ * Records errors thrown or rejected by Playwright reporters before Playwright converts them into run status.
+ *
+ * @param {object} reportersPackage
+ * @returns {object}
+ */
+function reportersHook (reportersPackage) {
+  return shimmer.wrap(reportersPackage, 'createReporters', createReporters => async function () {
+    const reporters = await createReporters.apply(this, arguments)
+    for (const reporter of reporters) {
+      if (instrumentedPlaywrightReporters.has(reporter) || typeof reporter.onEnd !== 'function') continue
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/playwright.js]: Fixed in [fix(playwright): cover remaining reporter failures](https://github.com/DataDog/dd-trace-js/commit/fbf5fe3f6). The per-reporter instrumentation now wraps both `onEnd` and `onExit` for Playwright 1.38–1.59. The exact Playwright 1.38 CI failure now passes locally.

File: packages/datadog-instrumentations/src/playwright.js [hunk 8]
```diff
@@ -1252,6 +1257,15 @@ function dispatcherHookNew (dispatcherExport, runWrapper) {
...
   // Config parameter is only available from >=1.55.0
   return async function (config) {
+    reporterError = undefined
+    hasReporterError = false
+    let restoreReporterConsoleError
+    if (satisfies(playwrightVersion, '>=1.60.0') && config?.config && !config.config[kDdPlaywrightReporterConfigured]) {
+      const DatadogPlaywrightReporter = require('./playwright-reporter')
+      Object.defineProperty(config.config, kDdPlaywrightReporterConfigured, { value: true })
+      config.config.reporter.unshift([require.resolve('./playwright-reporter')])
+      restoreReporterConsoleError = DatadogPlaywrightReporter.restoreConsoleError
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/playwright.js]: Fixed in [fix(playwright): cover remaining reporter failures](https://github.com/DataDog/dd-trace-js/commit/fbf5fe3f6). Console restoration is registered on every supported runner invocation while reporter insertion remains guarded. Added a real two-run regression using the same config object.

File: packages/datadog-instrumentations/src/playwright.js [hunk 9]
```diff
@@ -1550,6 +1644,26 @@ pageGotoCh.subscribe({
...
+ * @returns {void}
+ */
+function recordReporterError (error) {
+  if (hasReporterError) return
+
+  hasReporterError = true
+  const errorObject = error && typeof error === 'object' ? error : undefined
+  reporterError = new Error(errorObject?.message || errorObject?.value || String(error))
+  reporterError.name = errorObject?.name || reporterError.name
+  reporterError.stack = errorObject?.stack || reporterError.stack
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/playwright.js]: Fixed in [fix(playwright): cover remaining reporter failures](https://github.com/DataDog/dd-trace-js/commit/fbf5fe3f6). Reporter values are now normalized inside a guarded block before setting reporter-error state, with regression coverage for a hostile proxy whose metadata accessors throw.

**Comment by juan-fernandez** [comment]
@codex review

File: packages/datadog-instrumentations/src/playwright-reporter.js [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+   */
+  onEnd () {
+    this.isFinalizing = true
+    // Playwright 1.60 and 1.61 only expose reporter errors through this exact console call.
+    // eslint-disable-next-line no-console
+    const originalConsoleError = console.error
+    DatadogPlaywrightReporter.originalConsoleError = originalConsoleError
+    const reporter = this
+    // eslint-disable-next-line no-console
+    DatadogPlaywrightReporter.consoleError = console.error = function (message, error) {
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/playwright-reporter.js]: Fixed in fix(playwright): harden reporter finalization (https://github.com/DataDog/dd-trace-js/commit/ae426476afe8905e9313c626bdd324ff8fca17c6).

File: packages/datadog-instrumentations/src/playwright.js [hunk 11]
```diff
@@ -1367,7 +1383,27 @@ function runAllTestsWrapper (runAllTests, playwrightVersion) {
...
 
-    let runAllTestsReturn = await runAllTests.apply(this, arguments)
+    let runAllTestsReturn
+    try {
+      runAllTestsReturn = await runAllTests.apply(this, arguments)
+    } catch (error) {
+      try {
+        await getChannelPromise(testSessionFinishCh, {
+          status: 'fail',
+          error,
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/playwright.js]: Fixed in fix(playwright): harden reporter finalization (https://github.com/DataDog/dd-trace-js/commit/ae426476afe8905e9313c626bdd324ff8fca17c6).

File: packages/datadog-instrumentations/src/playwright-reporter.js [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+ * @returns {boolean}
+ */
+function isPlaywrightReporterError (message) {
+  if (message !== PLAYWRIGHT_REPORTER_ERROR_MESSAGE) return false
+
+  const originalPrepareStackTrace = Error.prepareStackTrace
+  try {
+    Error.prepareStackTrace = (_, callSites) => callSites
+    const stack = new Error('Playwright reporter error provenance').stack
+    return PLAYWRIGHT_REPORTER_ERROR_CALLER_RE.test(stack?.[2]?.toString() || '')
```
[the remark this answers is not in the document]
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/playwright-reporter.js]: Fixed in fix(playwright): harden reporter finalization (https://github.com/DataDog/dd-trace-js/commit/ae426476afe8905e9313c626bdd324ff8fca17c6).

**Comment by juan-fernandez** [comment]
@codex review

**Review by anmarchenko — APPROVED** [review]
